conary.1
.\" Copyright (c) 2004-2005 rPath, Inc.
.TH CONARY 1 "21 February 2005" "rPath, Inc."
.SH NAME
conary \- repository-based software configuration manager
.SH SYNOPSIS
.B conary \fIcommand \fR[\fI<args>\fR]
.SH DESCRIPTION
\fBConary\fR is a software configuration manager that uses distributed,
network-based, versioned repositories rather than disparate package
files as its canonical source of data.  Conary uses the repository
versioning to intelligently merge configuration files, file ownership
and permissions, and so forth.
.SH REFERENCE
Conary has several common command-line arguments, all of which modify 
configuration entries and are documented with other configuration
entries.  The rest of its command-line arguments are specific to
which conary \fBcommand\fP you are running, and are documented with
the commands. The \fIC\fPonary \fIv\fPersion \fIc\fPontrol program,
\fBcvc\fP, handles build and source management, and is documented in
cvc(1).
.SS "System Management"
The following \fBcommands\fP are used to install, update, erase, and
otherwise manage software installed on a local system.
.TP 4
.B changeset <\fItrovename\fP[[=\fIoldver\fP\fI[<flavor>]\fP>]--[\fInewver\fP\fI[<flavor>]]\fP>]+ \fIoutfile\fP
Creates an absolute changeset for \fItrovename newver\fP, or a relative
changeset between \fItrovename oldver\fP and \fItrovename newver\fP, and stores
it in outfile.
.RS 4
.TP
.B \-\-no-recurse
Don't recursively include troves referenced by the groups and packages
listed.  Normally referenced troves which do not match troves in the
\fBexcludeList\fR are included in the changeset.
.RE
.TP
.B commit \fIchangeset\fP [\-\-target-branch \fI<branch>\fP]
Commits a changeset to the repository.  This is not a reversible
operation, so take care.  The \fB\-\-target-branch\fP option
is only appropriate when you commit a \fBlocal changeset\fP
to a repository.
.TP
.B config  
Shows the configuration that conary will use for any 
subsequent commands executed, based on the configuration files it has 
loaded and the configuration-altering command-line options that were 
supplied.  See the \fIFILES\fP section of this manual for more information.
Arguments include
.RS 4
.TP 4
.B \-\-show-passwords
Normally, conary config will mask paswords in repository maps, to avoid their
accidental display.  You can override this masking with --show-passwords.  
.RE
.TP
.B emerge \fI<component>...\fP
Builds the recipe for the given component (which it finds by searching the
\fIinstallLabelPath\fP) onto the \fBlocalhost@local:EMERGE\fP branch
and updates the local system to the new build.
.TP
.B erase [+/-]\fItrovename\fP[=\fRversion][\fI<flavor>\fP]\fP+
Erase all instances of \fItrovename\fP (or only a specified \fIversion\fP)
from the local system. It has the same syntax as \fBupdate\fR, but trove
names which are not preceded by a plus or minus are removed instead of
updated.
.TP
.B localcommit
Applies a local changeset to the local system.  This is exactly
equivalent to making every change represented in the changeset
by hand on the local system, the locally-installed versions still
reference the repository as their source, future updates will still
come from the repository, and the changes represented in the local
changeset will be preserved as if they had been made by hand
locally.
.TP
.B localcs \fItrovename outfile\fP
Creates a local changeset, which represents the changes made on the
local system to \fItrovename\fP, relative to the package as it is
represented in the repository.
.TP
.B query \fItrovename\fP[=\fIversion\fP][\fI<flavor>\fP]*\fP [\fI\-\-path <path>\fP]*
Lists troves installed on the local system. Any number of trove names or
absolute paths may be given. Extra arguments may be given to change what
information is displayed (but they cannot be combined, unfortunately).
.RS 4
.TP 4
.B \-\-deps
Displays full dependency information
.TP
.B \-\-full-versions
Full version strings are displayed instead of abbreviated version numbers
.TP
.B \-\-info
Displays information on the specified troves, including flavors, requirements
and pin status.
.RE
.TP
.B rblist
Lists available rollbacks.
.TP
.B remove \fIpath\fP
Remove the file named \fIpath\fP from the system, and record in the
Conary database that the file has been removed.  After this, doing
\fBconary update\fP will not replace the file.  (If future versions
of the component rename the file, the removal will track the rename;
that is; if you run \fBconary remove /foo/bar\fP and \fI/foo/bar\fP
is later renamed \fI/foo/blah\fP, when you update to a version of
the component that now contains \fI/foo/blah\fP instead of
\fI/foo/bar\fP, \fI/foo/blah\fP will not be created on your system
as a new file.)
.TP
.B rollback \fIrollback\fP [\fIargs\fP] 
Applies rollback \fIrollback\fP.  (Syntax may change in future versions
of Conary.)
Arguments include
.RS 4
.TP 4
.B \-\-replace-files
See \-\-replace-files description from \fBconary update\fP
.RE
.TP
.B update [\fIargs\fP] [+/-]<\fItrovename\fP[=\fIversion\fP][\fI<flavor>\fP][--[\fIversion\fP][\fIflavor\fP]]>* <\fIchangeset-file\fP>*
Updates each \fItrovename\fR from the appropriate repository to 
match the optional \fIversion\fP and \fIflavor\fP given (otherwise the most 
recent version which matches the \fBinstall-label\fP is used).  Also 
applies any given changesets. 
If the \fItrovename\fR is preceded by a minus, the trove is erased as part
of the operation.  If a -- is used in a troveSpec, the version and flavor
before the -- (if any) specify the trove to erase from the local system,
and the version and flavor after the -- (if any) specify the update version.
.RS 4
.TP
.B \-\-exclude-troves \fIregexp-list\fR
Any recursively-included troves whose name matches one or more of the
regular expression in (the space separated) \fIregexp-list\fR will not be
newly installed. If they are already installed, they will be updated
normally.
.TP 4
.B \-\-from-file \fIchangeset-file\fR
Pick troves from \fIchangeset-file\fR, not from a networked repository.
.TP 4
.B \-\-keep-existing
When applying a changeset, do not remove any existing versions of
the troves included in the changeset.
This will work correctly only if the versions have entirely
disjoint file sets.
The \-\-keep-existing option is normally used to add another
branch with the same trove name to the system; after that,
each branch can be updated normally.
If two versions from the same branch are
installed at the same time, this may confuse future update
operations.
.TP
.B \-\-info
Display a summary of what troves will be affected by the update.
.TP
.B \-\-just-db
The update is performed on the database, but the filesystem is not
changed.
.TP
.B \-\-resolve
Attempt to automatically resolve all dependencies.
Can be made the default option by setting the autoResolve flag in conaryrc.
.TP
.B \-\-no-conflict-check
Skip check for troves being installed multiple times from a single branch.
.TP
.B \-\-no-deps
Skip dependency resolution and errors.
.TP
.B \-\-no-recurse
Install only the troves specified, skipping any troves they reference.
.TP
.B \-\-no-resolve
Do not resolve dependencies, only print out any dependency failures.
Unless \-\-resolve has been made the default by setting the
autoResolve flag in conaryrc, \-\-no-resolve
is the default behavior.
.TP
.B \-\-replace-files
When a normal file (not a configuration file or a transient file) has
changed on the system between updates to that file, or if a file
already exists when conary tries to create it, conary normally
aborts the operation with an error message.
If you want conary to update anyway, replacing the contents
of the file with the new version from the changeset, you will need
to provide the \-\-replace-files option.
.TP
.B \-\-sync
Only allow updates to versions that are referenced by other troves on the 
system.  This allows you to easily update a trove to a version specified 
within a group, or ensure you match the :lib component with an out-of-date
version of a trove you have installed.
.TP
.B \-\-test
Perform all actions up to the point of writing to the filesystem.
.RE
.TP
.B updateall
.\" FIXME: document
.RS 4
.TP 4
.B \-\-exclude-troves <patterns>
See \fBupdate \-\-exclude-troves\fP
.TP
.B \-\-info
Displays the list of update tasks that would be performed without
doing the actual system update
.TP
.B \-\-no-deps
See \fBupdate \-\-no-deps\fP
.TP
.B \-\-no-resolve
See \fBupdate \-\-no-replace\fP
.TP
.B \-\-replace-files
See \fBupdate \-\-replace-files\fP
.TP
.B \-\-resolve
See \fBupdate \-\-resolve\fP
.RE
.TP
.B update-conary
Downloads a stable version of the \fBconary client\fP and forcibly
installs it. This could help if the locally installed client is too
old (or otherwise misbehaves) and cannot perform a regular upgrade
through a \fBconary update conary\fP invocation.
.RE
.TP
.B repquery [\fIargs\fP] [\fItrovename\fP[=\fIversion\fP][\fI<flavor>\fP]]*
Lists troves in the repository.
Arguments include
.RS 4
.TP 4
.B \-\-all
.\" FIXME: document
.TP
.B \-\-buildreqs
Displays the troves that fulfilled the build requirements of the trove
.TP
.B \-\-deps
.\" FIXME: document
.TP
.B \-\-flavors
Displays flavors of the troves
.TP
.B \-\-full-versions
.\" FIXME: document
.TP
.B \-\-ids
.\" FIXME: document
.TP
.B \-\-info
.\" FIXME: document
.TP
.B \-\-leaves
.\" FIXME: document
.TP
.B \-\-ls
.\" FIXME: document
.TP
.B \-\-sha1s
.\" FIXME: document
.TP
.B \-\-tags
.\" FIXME: document
.RE
.RE
.TP
.B showcs [\fIargs\fP \fI<changeset> [<trove/component>=[<version>]*]\fP
The \fBshowcs\fP command lists the contents of the provided 
\fBchangeset\fP.  If a trove is specified, and that trove is within the 
changeset, only that trove is displayed.  A component specifier, such as 
:runtime, can be given instead of a trove.  In that case, all components 
within all packages in the changeset will be displayed.

.RS 4
.TP 4
.B \-\-full-versions
Does not attempt to shorten version strings for display
.\" FIXME: add semantics
.TP
.B \-\-deps
Displays provides and requires information about the troves in the changeset.
.TP
.B \-\-ls
Lists the files in the changeset
.TP
.B \-\-show-changes
For files that have been modified, lists the old version of the file above the
new version of the file for easy comparison.   Only has an effect if --ls
is specified.  
.TP
.B \-\-tags
Show any files in the changeset that have tags.
.RE
.TP
.B verify [--all] \fI<trove>[=version]*\fP
Compares the files in the given \fItrove\fP (or all troves if the --all 
option if given) against the trove files as they were  at the time of install,
and displays any differences.
.\"
.\"
.\"
.SH JARGON

NEWS
Changes in 0.70.4:
  o Client Changes
    * The trove that satisfies a dependency that is broken by erase is
      now displayed in the "Troves being removed create unresolved
      dependencies" message.
    * Components are now displayed on the same line as their parent
      package in "conary update" output.
    * A new 'interactive' option has been added to conary configuration. 
      When set to true, conary will display info about clone, branch, 
      update, and erase operations, and then ask before proceding.
  o Build Changes (cvc)
    * The CompilePython action has been fixed to accept macros at the
      beginning of its arguments, fixing a bug new in Conary 0.70.3.
    * The Requires policy can now be given synthetic RPATH elements;
      this is useful when programs are only intended to be run under
      scripts that set LD_LIBRARY_PATH and so do not intrinsically have
      the information they need to find their libraries.
  o Repository Changes
    * The repository now serves up stored OpenPGP keys as a "Limited
      Keyserver"; users can retrieve keys, but not search or browse them.
      The keys are available via /getOpenPGPKey?search=KEY_ID. This
      is meant only to allow conary to automatically retrieve OpenPGP
      keys used to sign packages.

Changes in 0.70.3:
  o Client Changes (conary)
    * Conary now works harder to avoid having separate erase/installs,
      instead preferring to link those up into one update when possible.
    * Conary configuration now supports contexts.  Contexts are defined in 
      sections starting with a [<name>] line, and provide contextual 
      configurations for certain variables, defined in the man page.  All 
      configuration options after the [<name>] will be associated with that 
      context, and will override the default configuration when that context 
      is active.  The current context can be selected by using the --context 
      parameter, or by setting the CONARY_CONTEXT environment variable. 
    * 'conary config --show-contexts' will display the available contexts
  o Build Changes (cvc)
    * A local cook of a trove foo will ensure that the changeset created is 
      installable on your local system, by making sure the version number 
      created is unique.
    * The builddir is no longer allowed to appear in ELF RPATHs.
    * The build documentation strings have been significantly updated
      to document the fact that for most strings, a relative path
      is relative to the builddir, but an absolute path is relative
      to the destdir.
    * The ManualConfigure action now sets the standard Configure
      environment.
    * cvc will allow you to cook a trove locally even when you are unable 
      to access the trove's source repository
  * Common Changes:
    * Version closeness was improperly measured for troves on different
      branches when then label structure was identical
  o Repository Changes
    * Repository now has a config flag called requireSigs. Setting it to
      True will force all troves to have valid package signatures.  Troves
      lacking this will be rejected.  Enabling this option prevents the
      generation of branches, shadows, or clones since these troves are not
      signed.  It is not recommended that this option be enabled until the
      infrastructure is in place to provide package signatures for all types
      of troves.

Changes in 0.70.2:
  o Client Changes (conary)
    * GnuPG compatible trust metrics for OpenPGP Keys now exists. This
      makes it possible for conary clients to refuse troves that
      aren't properly trusted. The metrics currently in place mimic
      gpg behavior.
    * Running "conary update" in a directory that does not exist no
      longer fails with an error (bugzilla #212).  Note that "cvc
      update" still requires that the current working directory exists
      of course.
    * HTTP error conditions are handled more gracefully when commiting
      a change set. (bugzilla #334)
    * conary more reliably sets a non-zero exit status when an error
      occurs. (bugzilla #312)
    * When performing an update of a group that adds a trove foo,
      search the system for a older version of foo to replace if the
      original update command found a replacement by searching the
      system.
    * New option, "conary update-conary" has been added in an attempt
      to provide a workaround for future drastic protocol revisions
      such as what happened for 0.70 
    * Methods for parsing command line update request and changeset requests
      have been added to conaryclient.cmdline
    * A metric for the distance between arbitrary versions on different
      branches has been added, and the code which matches troves changes
      between collections uses this code to give well-defined matches
      for all cases.
    * Rollbacks are now listed with the most recent on top
    * Troves which a group operation tries to remove will be left behind
      if they satisfy dependencies for other troves
    * updateall command respects pins on top-level troves
    * Dependency resolution no longer blows away pinned troves
    * conary update now takes a changeSpec, allowing you to specify both

conary.py
#!/usr/bin/python
# -*- mode: python -*-
#
# Copyright (c) 2004-2005 rPath, Inc.
#
# This program is distributed under the terms of the Common Public License,
# version 1.0. A copy of this license should have been distributed with this
# source file in a file called LICENSE. If it is not present, the license
# is always available at http://www.opensource.org/licenses/cpl.php.
#
# This program is distributed in the hope that it will be useful, but
# without any waranty; without even the implied warranty of merchantability
# or fitness for a particular purpose. See the Common Public License for
# full details.
#
"""
The conary main program.
"""

import sys
if sys.version_info < (2, 4):
    print "error: python 2.4 or later is requried"
    sys.exit(1)

#stdlib
import os
import xmlrpclib

#conary
import callbacks
import checkin
import commit
import conarycfg
import constants
import cscmd
import deps
import display
import flavorcfg
from lib import log
from lib import options
from lib import util
from local import database
import queryrep
import conaryclient
import rollbacks
import showchangeset
import updatecmd
import verify
import versions
from trove import DigitalSignatureVerificationError

sys.excepthook = util.genExcepthook()

def usage(rc = 1):
    print "usage: conary changeset <pkg>[=[<oldver>--]<newver>]+ <outfile>"
    print "       conary commit       <changeset>"
    print "       conary config"
    print "       conary emerge       <troveName>+"
    print "       conary erase        <pkgname>[=<version>][[flavor]]+"
    print "       conary localcs      <pkg> <outfile>"
    print "       conary localcommit  <changeset>"
    print "       conary pin          <pkgname>[=<version>][[flavor]]*"
    print "       conary query        <pkgname>[=<version>][[flavor]]*"
    print "       conary remove       <path>"
    print "       conary repquery     <pkgname>[=<version>][[flavor]]*"
    print "       conary rblist"
    print "       conary rollback     <rollback>"
    print "       conary showcs       <changeset> <trove>[=<version>]*"
    print "       conary unpin        <pkgname>[=<version>][[flavor]]*"
    print "       conary update       <pkgname>[=<version>][[flavor]]* <changeset>*"
    print "       conary updateall"
    print "       conary update-conary"
    print "       conary usage"
    print "       conary verify       <pkgname>[=<version>][[flavor]]*"
    print "       conary --version"
    print ""
    print "changeset flags: --exclude-troves <patterns>"
    print "                 --no-recurse"
    print "                 --quiet"
    print ""
    print "commit flags:    --target-branch <branch>"
    print ""
    print "config flags:    --show-passwords --show-contexts"
    print ""
    print 'common flags:    --build-label <label>'
    print '                 --config-file <path>'
    print '                 --config "<item> <value>"'
    print '                 --install-label <label>'
    print "                 --root <root>"
    print "                 --context <context>"
    print ""
    print "query flags:     --buildreqs"
    print "                 --deps"
    print "                 --diff"
    print "                 --flavors"
    print "                 --full-versions"
    print "                 --ids"
    print "                 --info"
    print "                 --ls"
    print "                 --path <file>"
    print "                 --sha1s"
    print "                 --tags"
    print ""
    print "repquery flags:  --all"
    print "                 --buildreqs"    
    print "                 --deps"    
    print "                 --full-versions"
    print "                 --flavors"
    print "                 --ids"
    print "                 --info"
    print "                 --leaves"
    print "                 --ls"
    print "                 --sha1s"
    print "                 --tags"
    print
    print "rollback flags:  --replace-files"
    print
    print "showcs flags:    --full-versions"
    print "                 --info"
    print "                 --ls"
    print "                 --show-changes"
    print "                 --tags"
    print ""
    print "update/erase flags:"
    print "                 --exclude-troves <patterns>"
    print "                 --from-file <file.ccs>"
    print "                 --info"
    print "                 --just-db"
    print "                 --keep-existing"
    print "                 --no-conflict-check"
    print "                 --no-deps"
    print "                 --no-recurse"
    print "                 --no-resolve"
    print "                 --quiet"
    print "                 --replace-files"
    print "                 --resolve"
    print "                 --sync"
    print "                 --test"
    print "updateall flags:"
    print "                 --exclude-troves <patterns>"
    print "                 --info"
    print "                 --no-deps"
    print "                 --no-resolve"
    print "                 --replace-files"
    print "                 --resolve"
    return rc

def openRepository(repMap):
    from repository import netclient, errors
    try:
        return netclient.NetworkRepositoryClient(repMap)
    except errors.OpenError, e:
	log.error('Unable to open repository %s: %s', path, str(e))
	sys.exit(1)

def openDatabase(root, path):
    return database.Database(root, path)

def realMain(cfg, argv=sys.argv):
    argDef = {}
    cfgMap = {}

    cfgMap["build-label"] = "buildLabel"
    cfgMap["exclude-troves"] = "excludeTroves"
    cfgMap["root"] = "root"
    cfgMap["trust-threshold"] = "trustThreshold"

    (NO_PARAM,  ONE_PARAM)  = (options.NO_PARAM, options.ONE_PARAM)
    (OPT_PARAM, MULT_PARAM) = (options.OPT_PARAM, options.MULT_PARAM)

    argDef["all"] = NO_PARAM
    argDef["buildreqs"] = NO_PARAM
    argDef["config"] = MULT_PARAM
    argDef["config-file"] = ONE_PARAM
    argDef["context"] = ONE_PARAM
    argDef["deps"] = NO_PARAM
    argDef["diff"] = NO_PARAM
    argDef["from-file"] = MULT_PARAM
    argDef["flavors"] = NO_PARAM
    argDef["full-versions"] = NO_PARAM
    argDef["ids"] = NO_PARAM
    argDef["info"] = NO_PARAM
    argDef["install-label"] = MULT_PARAM
    argDef["items"] = NO_PARAM
    argDef["just-db"] = NO_PARAM
    argDef["keep-existing"] = NO_PARAM
    argDef["no-deps"] = NO_PARAM
    argDef["no-recurse"] = NO_PARAM
    argDef["resolve"] = NO_PARAM
    argDef["no-resolve"] = NO_PARAM
    argDef["no-conflict-check"] = NO_PARAM
    argDef["leaves"] = NO_PARAM
    argDef["path"] = MULT_PARAM
    argDef["ls"] = NO_PARAM
    argDef["profile"] = NO_PARAM
    argDef["quiet"] = NO_PARAM
    argDef["replace-files"] = NO_PARAM
    argDef["sha1s"] = NO_PARAM
    argDef["show-changes"] = NO_PARAM
    argDef["sync"] = NO_PARAM
    argDef["tag-script"] = ONE_PARAM
    argDef["tags"] = NO_PARAM
    argDef["target-branch"] = ONE_PARAM
    argDef["test"] = NO_PARAM
    argDef["version"] = NO_PARAM
    argDef["show-passwords"] = NO_PARAM
    argDef["show-contexts"] = NO_PARAM

    try:
        argSet, otherArgs = options.processArgs(argDef, cfgMap, cfg, usage,
                                                argv=argv)
    except options.OptionError, e:
        print >> sys.stderr, e
        sys.exit(e.val)
    except versions.ParseError, e:
	print >> sys.stderr, e
	sys.exit(1)

    if argSet.has_key('version'):
        print constants.version
        sys.exit(0)

    context = cfg.context
    if os.path.exists('CONARY'):
        state = checkin.ConaryStateFromFile('CONARY')
        if state.hasContext():
            context = state.getContext()

    context = os.environ.get('CONARY_CONTEXT', context)
    context = argSet.pop('context', context)

    if context:
        cfg.setContext(context)

    l = []
    for labelStr in argSet.get('install-label', []):
        l.append(versions.Label(labelStr))
    if l:
        cfg.installLabelPath = l
        del argSet['install-label']

    if cfg.installLabelPath:
        cfg.installLabel = cfg.installLabelPath[0]
    
    cfg.initializeFlavors()

    profile = False
    if argSet.has_key('profile'):
	import hotshot
	prof = hotshot.Profile('conary.prof')
	prof.start()
	profile = True
	del argSet['profile']

    if (len(otherArgs) < 2):
	return usage()
    elif (otherArgs[1] == "changeset"):
        kwargs = {}
        
        callback = updatecmd.UpdateCallback()
        if cfg.quiet:
            callback = callbacks.UpdateCallback()
        if argSet.has_key('quiet'):
            callback = callbacks.UpdateCallback()
            del argSet['quiet']
        kwargs['callback'] = callback

        kwargs['recurse'] = not(argSet.has_key('no-recurse'))
        if not kwargs['recurse']:
            del argSet['no-recurse']
            
	if len(otherArgs) < 4 or argSet:
	    return usage()

        outFile = otherArgs[-1]
        del otherArgs[-1]

	repos = openRepository(cfg.repositoryMap)

	cscmd.ChangeSetCommand(repos, cfg, otherArgs[2:], outFile, **kwargs)
    elif (otherArgs[1] == "commit"):
	targetBranch = None
	if argSet.has_key('target-branch'):
	    targetBranch  = argSet['target-branch']
	    del argSet['target-branch']
	if len(otherArgs) < 3: return usage()
	repos = openRepository(cfg.repositoryMap)
	for changeSet in otherArgs[2:]:
	    commit.doCommit(repos, changeSet, targetBranch)
    elif (otherArgs[1] == "config"):
        
	showPasswords = argSet.pop('show-passwords', False)
	showContexts = argSet.pop('show-contexts', False)
        try:
            prettyPrint = sys.stdout.isatty()
        except AttributeError:
            prettyPrint = False
        cfg.setDisplayOptions(hidePasswords=not showPasswords,
                              showContexts=showContexts,
                              prettyPrint=prettyPrint)
	if argSet: return usage()
	if (len(otherArgs) > 2):
	    return usage()
	else:
	    cfg.display()
    elif (otherArgs[1] == "emerge"):
        # import this late to reduce the dependency set for
        # the main conary command in the common case.  This lets
        # conary run even if, for example, libelf is missing
        from build import cook
	log.setVerbosity(log.DEBUG)

	if argSet: return usage()
        # XXX this is pretty broken, there should be a default in the cfg
        #     object for noClean
        cfg.noClean = False
	cook.cookCommand(cfg, otherArgs[2:], False, {}, emerge = True)
    elif (otherArgs[1] == "localcs"):
	if len(otherArgs) != 4 and len(otherArgs) != 4:
	    return usage()

	name = otherArgs[2]
	outFile = otherArgs[3]

	db = database.Database(cfg.root, cfg.dbPath)
	cscmd.LocalChangeSetCommand(db, cfg, name, outFile)
    elif (otherArgs[1] == "localcommit"):
	if len(otherArgs) < 3: return usage()
	db = database.Database(cfg.root, cfg.dbPath)
	for changeSet in otherArgs[2:]:
	    commit.doLocalCommit(db, changeSet)
    elif (otherArgs[1] == "pin" or otherArgs[1] == "unpin"):
	if argSet: return usage()

        updatecmd.changePins(cfg, otherArgs[2:], pin = otherArgs[1] == "pin")
    elif (otherArgs[1] == "query") or (otherArgs[1] == "q"):
        paths = argSet.pop('path', [])
	tags = argSet.pop('tags', False)
        info = argSet.pop('info', False)
	ls = argSet.pop('ls', False)
	deps = argSet.pop('deps', False)
	showDiff = argSet.pop('diff', False)
	ids = argSet.pop('ids', False)
	sha1s = argSet.pop('sha1s', False)
	fullVersions = argSet.pop('full-versions', False)
	showFlavors = argSet.pop('flavors', False)
	showBuildReqs = argSet.pop('buildreqs', False)

	db = openDatabase(cfg.root, cfg.dbPath)

	if argSet: return usage()

	if len(otherArgs) >= 2:
	    try:
                display.displayTroves(db, otherArgs[2:], paths, ls, ids, sha1s,
                                      fullVersions, tags, info=info, deps=deps,
                                      showBuildReqs=showBuildReqs, 
                                      showFlavors=showFlavors, showDiff=showDiff)
	    except IOError, msg:
		sys.stderr.write(msg.strerror + '\n')
		return 1
	else:
	    return usage()
    elif (otherArgs[1] == "repquery") or (otherArgs[1] == "rq"):
        cfg.requireInstallLabelPath()
	all = argSet.pop('all', False)
	ls = argSet.pop('ls', False)
	fullVersions = argSet.pop('full-versions', False)
	ids = argSet.pop('ids', False)
	info = argSet.pop('info', False)
	tags = argSet.pop('tags', False)
	sha1s = argSet.pop('sha1s', False)
	leaves = argSet.pop('leaves', False)
	showDeps = argSet.pop('deps', False)
	showBuildReqs = argSet.pop('buildreqs', False)
	showFlavors = argSet.pop('flavors', False)

	repos = openRepository(cfg.repositoryMap)

	if argSet: return usage()

	if len(otherArgs) >= 2:
	    args = [repos, cfg, otherArgs[2:], all, ls, ids, sha1s, leaves, 
                    fullVersions, info, tags, showDeps, showBuildReqs,
                    showFlavors]
	    try:
		queryrep.displayTroves(*args)
	    except IOError, msg:
                # XXX when is a msg.strerror not a str?
                # at least socket.gaierror, which is a tuple of
                # return code and string
		sys.stderr.write(str(msg.strerror) + '\n')
		sys.exit(1)
	else:
	    return usage()
    elif (otherArgs[1] == "rblist"):
	if argSet: return usage()
	db = openDatabase(cfg.root, cfg.dbPath)
	rollbacks.listRollbacks(db, cfg)
    elif (otherArgs[1] == "remove"):
	if len(otherArgs) != 3: return usage()
	if argSet: return usage()
	db = openDatabase(cfg.root, cfg.dbPath)
	fullPath = util.joinPaths(cfg.root, otherArgs[2])
	if os.path.exists(fullPath):
	    os.unlink(fullPath)
	else:
	    log.warning("%s has already been removed", fullPath)
	db.removeFile(otherArgs[2])
    elif (otherArgs[1] == "rollback"):
        kwargs = {}
	if argSet.has_key('replace-files'):
	    kwargs['replaceFiles'] = True
	    del argSet['replace-files']
	if argSet: return usage()
	repos = openRepository(cfg.repositoryMap)
	db = openDatabase(cfg.root, cfg.dbPath)
	args = [db, repos, cfg] + otherArgs[2:]
	rollbacks.apply(*args, **kwargs)
    elif (otherArgs[1] == "verify"):
	db = openDatabase(cfg.root, cfg.dbPath)
        all = argSet.has_key('all')
	if all: del argSet['all']
        if len(otherArgs) < 2 or argSet:
            return verify.usage()
        troves = otherArgs[2:]
        verify.verify(troves, db, cfg, all=all)
    elif (otherArgs[1] == "showcs" or otherArgs[1] == "scs"):
        from repository import changeset
        ls = argSet.has_key('ls')
	if ls: del argSet['ls']

        all = argSet.has_key('all')
	if all: del argSet['all']

        tags = argSet.has_key('tags')
	if tags: del argSet['tags']

        sha1s = argSet.has_key('sha1s')
	if sha1s: del argSet['sha1s']

        ids = argSet.has_key('ids')
	if ids: del argSet['ids']

        info = argSet.has_key('info')
	if info: del argSet['info']

        showDeps = argSet.has_key('deps')
	if showDeps: del argSet['deps']

        showChanges = argSet.has_key('show-changes')
	if showChanges: del argSet['show-changes']

        fullVersions = argSet.has_key('full-versions')
	if fullVersions: del argSet['full-versions']

        if argSet: return showchangeset.usage()

        if len(otherArgs) < 3:
            showchangeset.usage()
            return 1
        cspath = otherArgs[2]
        component = None
        if len(otherArgs) > 3:
            component = otherArgs[3:]
        cs = changeset.ChangeSetFromFile(cspath)
	db = database.Database(cfg.root, cfg.dbPath)
	repos = openRepository(cfg.repositoryMap)
        showchangeset.displayChangeSet(db, repos, cs, component, cfg, ls, 
                                        tags, fullVersions, showChanges, 
                                        ids=ids, sha1s=sha1s, all=all, 
                                        deps=showDeps)
    elif otherArgs[1] == "update-conary":
        cfg.requireInstallLabelPath()
        updatecmd.updateConary(cfg, constants.version)
    elif otherArgs[1] in ["update", "erase"]:
        cfg.requireInstallLabelPath()
	kwargs = {}

        callback = updatecmd.UpdateCallback()
        if cfg.quiet:
            callback = callbacks.UpdateCallback()
        if argSet.has_key('quiet'):
            callback = callbacks.UpdateCallback()
            del argSet['quiet']
        kwargs['callback'] = callback

	if argSet.has_key('resolve'):
            cfg.autoResolve = True
	    del argSet['resolve']

	if argSet.has_key('no-resolve'):
            cfg.autoResolve = False
	    del argSet['no-resolve']

        kwargs['replaceFiles'] = argSet.pop('replace-files', False)
        kwargs['depCheck'] = not argSet.pop('no-deps', False)
        kwargs['fromFiles'] = argSet.pop('from-file', [])
        kwargs['recurse'] = not argSet.pop('no-recurse', False)
        kwargs['checkPathConflicts'] = \
                                not argSet.pop('no-conflict-check', False)
        kwargs['justDatabase'] = argSet.pop('just-db', False)
        kwargs['info'] = argSet.pop('info', False)
        kwargs['keepExisting'] = argSet.pop('keep-existing', False)
        kwargs['tagScript'] = argSet.pop('tag-script', None)
        kwargs['test'] = argSet.pop('test', False)
        kwargs['sync'] = argSet.pop('sync', False)
        kwargs['updateByDefault'] = (otherArgs[1] == "update")

        if kwargs['sync'] and kwargs['fromFiles']:
            log.error("Only one of --sync and --from-file may be used")
            return 1

	if argSet: return usage()
	if len(otherArgs) >=3:
	    updatecmd.doUpdate(cfg, otherArgs[2:], **kwargs)
	else:
	    return usage()
    elif (otherArgs[1] == "updateall"):
        cfg.requireInstallLabelPath()
	kwargs = {}

	if argSet.has_key('info'):
	    kwargs['info'] = True
	    del argSet['info']

	if argSet.has_key('items'):
	    kwargs['showItems'] = True
	    del argSet['items']

	if argSet.has_key('no-deps'):
	    kwargs['depCheck'] = False
	    del argSet['no-deps']

	if argSet.has_key('replace-files'):
	    kwargs['replaceFiles'] = True
	    del argSet['replace-files']

	if argSet.has_key('no-resolve'):
            cfg.autoResolve = False
	    del argSet['no-resolve']

	if argSet.has_key('resolve'):
            cfg.autoResolve = True
	    del argSet['resolve']

	if argSet.has_key('test'):
	    kwargs['test'] = argSet['test']
	    del argSet['test']

	if argSet: return usage()

	if len(otherArgs) == 2:
	    updatecmd.updateAll(cfg, **kwargs)
	else:
	    return usage()
    elif (otherArgs[1] == "return usage"):
	return usage(rc = 0)
    else:
	return usage()

    if profile:
	prof.stop()

    if log.errorOccurred():
	sys.exit(1)

def main(argv=sys.argv):
    from repository import errors
    try:
        if '--skip-default-config' in argv:
            argv = argv[:]
            argv.remove('--skip-default-config')
            cfg = conarycfg.ConaryConfiguration(False)
        else:
            cfg = conarycfg.ConaryConfiguration()

        # reset the excepthook (using cfg values for exception settings)
        sys.excepthook = util.genExcepthook(cfg.dumpStackOnError)
	return realMain(cfg, argv)
    except conarycfg.ConaryCfgError, e:
        log.error(str(e))
    except xmlrpclib.ProtocolError, e:
	if e.errcode == 403:
	    print >> sys.stderr, \
		"remote server denied permission for the requested operation"
	else:
	    raise
    except errors.UnknownException, e:
	print >> sys.stderr, \
	    "An unknown exception occured on the repository server:"
	print >> sys.stderr, "\t%s" % str(e)
    except errors.RepositoryError, e:
	print >> sys.stderr, str(e)
    except database.OpenError, e:
	print >> sys.stderr, str(e)
    except conaryclient.cmdline.TroveSpecError, e:
	print >> sys.stderr, str(e)
    except database.OldDatabaseSchema, e:
	print >> sys.stderr, str(e)
    except conaryclient.UpdateError, e:
        print >> sys.stderr, str(e)
    except conaryclient.CloneError, e:
        print >> sys.stderr, str(e)
    except conaryclient.InstallPathConflicts, e:
        print >> sys.stderr, str(e)
    except DigitalSignatureVerificationError, e:
        print >> sys.stderr, str(e)
    except:
        raise
    return 1

if __name__ == "__main__":
    sys.exit(main())
